### ISSUE TITLE
Switch to using keywords for syntax highlighting

### ISSUE THREAD

**Issue by petdance** [vim-perl/vim-perl#277]
Prompted by @roflcopter4's PR #276, explore using keywords for syntax highlighting.

Example: Replace

```
syn match perlStatementIPC "\v<%(msg%(ctl|get|rcv|snd)|sem%(ctl|get|op)|shm%(ctl|get|read|write))>"
```

with

```
syn keyword perlStatementIPC msgctl msgget msgrcv msgsnd semctl semget semop shmctl shmget shmread shmwrite
```

It's certainly easier to read. Is it faster? Can we measure?

**Comment by roflcopter4** [comment]
I was a bit concerned in this sort of case. It's conceivable that regular expressions may save time for groups of keywords like this that are so similar. I'm not sure how you'd test this though. I've only really noticed the difference between match and keyword in a few canned cases that involved flooding vim with tens of thousands of things to check.

**Comment by hoelzro** [comment]
Thanks again for looking into this, @roflcopter4.  AFAIK the only reason for "correctness" that impacts the use of keyword vs match is that the fat comma operator won't quote its left-hand side in the case of keywords - for example:

Match (`push` is highlighted as a string):
![match](https://user-images.githubusercontent.com/107804/60379805-028cc180-99ff-11e9-8ed0-7acd68bd9fec.png)

Keyword (`push` is highlighted as a keyword):
![keyword](https://user-images.githubusercontent.com/107804/60379807-07517580-99ff-11e9-811b-559bb005f790.png)

As far as performance goes, I measured last year and saw a negligible difference - sadly, while I documented _that_, I didn't write down what version of Vim I was using, or what files I was testing against.  Out of all likelihood I was testing the highlighting of the perl5 corpus that the vim-perl tests run against.  If you're interested, I can dig up this information.

Regardless of the files I've tested against, I have a feeling the corpus doesn't contain some of the more pathological cases that people run into with vim-perl, so I wouldn't be suprised if keywords helped more in larger files!

**Comment by petdance** [comment]
I'm concerned about the fat comma issue. I think that having `push => 3` highlighting "push" as a keyword would be a serious step backwards.

**Comment by hoelzro** [comment]
The best thing I can think of is to make it an option: some people might be willing to sacrifice the correct highlighting in favor of better speed.  I'd want to see some benchmarks that show substantial improvements in some slower-to-highlight files to justify that, though.

Another option, while it's very new and experimental, is to play with Vim's new [text properties](https://vimhelp.org/textprop.txt.html) feature.  Even once that API stabilizes, we would probably have to support `syn match`-based keywords for a while, so it might not be worth the additional complexity.

**Comment by roflcopter4** [comment]
I've run into this sort of issue before in my various code highlighting plugins. Every time it's reared its ugly head I have failed to solve it in any kind of satisfactory way. The only option that seems to work at all reliably is to use syntax matches instead of keywords. I didn't think this would crop up here. Looks like I forgot that this is Perl we're talking about.

I'm not even going to try to solve this. I've tried before, believe me. Maybe using a new feature like that could be possible (does Neovim support it too?) at some point but who knows. For now the smart thing is just to give up.
